src/Microsoft.IdentityModel.JsonWebTokens/JsonWebToken.cs

```diff
@@ -513,11 +537,192 @@ internal void ReadToken(string encodedJson)
+        /// <exception cref="SecurityTokenMalformedException">if <paramref name="encodedJsonSpan"/> has more than 4 dots.</exception>
+        internal void ReadToken(ReadOnlySpan<char> encodedJsonSpan)
+        {
+            // JWT must have 2 dots for JWS or 4 dots for JWE (a.b.c.d.e)
+            ReadOnlySpan<char> encodedJsonSpanCopy = encodedJsonSpan;
+
+            int dotIndex = encodedJsonSpanCopy.IndexOf("."); 
+            if (dotIndex == -1 || dotIndex == encodedJsonSpan.Length - 1)
+                throw LogHelper.LogExceptionMessage(new SecurityTokenMalformedException(LogMessages.IDX14100));
+
+            // segment before first dot (a)
+            ReadOnlySpan<char> headerSpan = encodedJsonSpanCopy.Slice(0, dotIndex);
+            Dot1 = dotIndex;
+
+            // b.c.d.e
+            encodedJsonSpanCopy = encodedJsonSpanCopy.Slice(dotIndex + 1);
+            dotIndex = encodedJsonSpanCopy.IndexOf(".");
+            if (dotIndex == -1)
+                throw LogHelper.LogExceptionMessage(new SecurityTokenMalformedException(LogMessages.IDX14120));
+
+            // segment between first and second dot, can be a payload for JWS or an encrypted key for JWE (b)
+            ReadOnlySpan<char> secondSpan = encodedJsonSpanCopy.Slice(0, dotIndex);
+            Dot2 = Dot1 + dotIndex + 1;
+
+            if (dotIndex == encodedJsonSpanCopy.Length - 1)
+            {
+                dotIndex = -1;
+            }
+            else
+            {
+                // c.d.e
+                encodedJsonSpanCopy = encodedJsonSpanCopy.Slice(dotIndex + 1);
+                dotIndex = encodedJsonSpanCopy.IndexOf('.'); // third dot
+            }
+
+            if (dotIndex == -1)
+            {
+                // JWS has two dots
+                // JWS: https://www.rfc-editor.org/rfc/rfc7515
+                // Format: https://www.rfc-editor.org/rfc/rfc7515#page-7
+
+                // add length of header and second span, plus 2 dots
+                IsSigned = !(headerSpan.Length + secondSpan.Length + 2 == encodedJsonSpan.Length);
+                try
+                {
+                    Header = CreateClaimSet(headerSpan, 0, headerSpan.Length, CreateHeaderClaimSet);
+                }
+                catch (Exception ex)
+                {
+                    throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(
+                        LogMessages.IDX14102,
+                        LogHelper.MarkAsUnsafeSecurityArtifact(headerSpan.ToString(), t => t.ToString())), // TODO: Add an overload to LogHelper.MarkAsUnsafeSecurityArtifact that accepts span?
+                        ex));
+                }
+
+                try
+                {
+                    Payload = CreateClaimSet(secondSpan, 0, secondSpan.Length, CreatePayloadClaimSet);
```

Comment:
We should add another ctor to https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/e7eb8b717aae2001ee4d40fc974adca6a33c4445/src/Microsoft.IdentityModel.JsonWebTokens/Json/JsonWebToken.PayloadClaimSet.cs#L15 that takes a span.

Reply:
Span cannot be used as an argument to a generic type. See
[this section by Adam Sitnik](https://adamsitnik.com/Span/#span-must-not-be-a-generic-type-argument).

Reply:
Looking into SpanAction and how it can be used.

Reply:
I don't understand. Why can't CreatePayloadClaimSet take a Span<byte>?

Reply:
CreatePayloadClaimSet is an action that is performed on the decoded string. 
When you see [here](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blame/0b783fbda1c047c0905186a8136c407999389824/src/Microsoft.IdentityModel.JsonWebTokens/JsonWebToken.cs#L516) this action is of the form is a Func<byte[], int, JsonClaimSet>.
Even though, an overload can be added for CreatePayloadClaimSet to accept a span, it cannot be invoked by using <Func<ReadOnlySpan<byte>, int, JsonClaimSet>. It leads to a compilation error.

An alternative would be to invoke the method directly without using Func.

Reply:
thanks for explaining, I think we need to refactor in a way we can pass the span.

Reply:
Refactored CreateClaimSet and updated CreateHeaderClaimSet and CreatePayloadClaimSet to accept spans